device/cuda/include/traccc/cuda/clusterization/clusterization_algorithm.hpp

```diff
@@ -67,12 +78,31 @@ class clusterization_algorithm
     /// @param det_descr The detector description
     /// @return a measurement collection (buffer)
     ///
-    output_type operator()(
+    /// @{
+    measurement_collection_types::buffer operator()(
         const edm::silicon_cell_collection::const_view& cells,
         const silicon_detector_description::const_view& det_descr)
         const override;
 
+    measurement_collection_types::buffer operator()(
+        const edm::silicon_cell_collection::const_view& cells,
+        const silicon_detector_description::const_view& det_descr,
+        device::clustering_discard_disjoint_set&&) const override;
+
+    std::pair<measurement_collection_types::buffer,
+              vecmem::data::vector_buffer<unsigned int>>
+    operator()(const edm::silicon_cell_collection::const_view& cells,
+               const silicon_detector_description::const_view& det_descr,
+               device::clustering_keep_disjoint_set&&) const override;
```

Comment:
Using a `struct` like this is an interesting idea. I would not have gone this way, but I kind of like it. As it results in pretty nicely readable client code.

However the return value **must** be:

```c++
std::pair<traccc::measurement_collection_types::buffer, traccc::edm::silicon_cluster_collection::buffer>
```

The algorithm itself must take care of creating the cluster EDM, not the client code.

Reply:
Sure, we can do that. :+1: